src/typeInference/README.md

```diff
@@ -0,0 +1,39 @@
+# Type Inference
+
+## Annotation
+
+The types of each node in the AST need to be determined to be used for optimisation purposes.
+
+Annotation starts by enabling the `annotateAST` flags in the options parameter of the `evaluateXPath` functions. When the flag is enabled, the `annotateAST` function gets triggered, recursively performs a depth-first, pre-order traversal to annotate every subtree of the AST. Given the type of the sub-tree, the corresponding annotation function gets called. These functions perform the annotation by first determining the type and then inserting a `SequenceType` to the AST as an attribute called `type`.
+
+-   Function references are being resolved by first extracting the `QName` out of the AST and then performing a lookup in the `StaticContext` (passed as an optional attribute of the `AnnotationContext` class) to determine the arity, function name and the return type of the function.
```

Comment:
Are you also already resolving namespace URIs? That could also be cool: for any prefix in the AST, also resolve the namespaceURI: convert `self::docbook:para` to effectively `self::Q{http://docbook.org/ns/docbook}para`. This will remove a bit of work from other steps of AST / compilation. I think this does not need to be in the PR, but could be a very cool future improvement.